lib/rubygems/update_suggestion.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+##
+# Mixin methods for Gem::Command to promote available RubyGems update
+
+module Gem::UpdateSuggestion
+  # list taken from https://github.com/watson/ci-info/blob/7a3c30d/index.js#L56-L66
+  CI_ENV_VARS = [
+    "CI", # Travis CI, CircleCI, Cirrus CI, Gitlab CI, Appveyor, CodeShip, dsari
+    "CONTINUOUS_INTEGRATION", # Travis CI, Cirrus CI
+    "BUILD_NUMBER", # Jenkins, TeamCity
+    "CI_APP_ID", "CI_BUILD_ID", "CI_BUILD_NUMBER", # Applfow
+    "RUN_ID" # TaskCluster, dsari
+  ].freeze
+
+  ONE_WEEK = 7 * 24 * 60 * 60
+
+  ##
+  # Message to promote available RubyGems update with related gem update command.
+
+  def update_suggestion
+    <<-MESSAGE
+
+A new release of RubyGems is available: #{Gem.rubygems_version} → #{Gem.latest_rubygems_version}!
+Run `gem update --system #{Gem.latest_rubygems_version}` to update your installation.
+
+    MESSAGE
+  end
+
+  ##
+  # Determines if current environment is eglible for update suggestion.
+
+  def eglible_for_update?
+    # explicit opt-out
+    return false if Gem.configuration[:prevent_update_suggestion]
+    return false if ENV["RUBYGEMS_PREVENT_UPDATE_SUGGESTION"]
+
+    # focus only on human usage of final RubyGems releases
+    return false unless Gem.ui.tty?
+    return false if Gem.rubygems_version.prerelease?
+    return false if Gem.disable_system_update_message
+    return false if ci?
+
+    # check makes sense only when we can store of last try
+    # otherwise we will not be able to prevent annoying update message
+    # on each command call
+    return unless Gem.configuration.config_file_writable?
+
+    # load time of last check, ensure the difference is enough to repeat the suggestion
+    check_time = Time.now.to_i
+    last_update_check = Gem.configuration[:last_update_check] || 0
+    return false if (check_time - last_update_check) < ONE_WEEK
+
+    # compare current and latest version, this is the part where
+    # latest rubygems spec is fetched from remote
+    (Gem.rubygems_version < Gem.latest_rubygems_version).tap do |eglible|
+      if eglible
+        # store the time of last successful check into config file
+        Gem.configuration[:last_update_check] = check_time
+        Gem.configuration.write
```

Comment:
Would it be possible to store this somewhere other than the configuration file? 

Libraries like the `update-notifier` npm package write to a file in `~/.config` directory, which is undesired because people version control it. I [opened an issue there](https://github.com/yeoman/update-notifier/issues/180) and the maintainer suggested `XDG_DATA_HOME` would be better, [npm uses a cache directory](https://github.com/npm/cli/blob/e95017a07b041cbb3293e659dad853f76462c108/lib/utils/update-notifier.js#L27)

Reply:
I believe `Gem.configuration` already respects `XDG_DATA_HOME` if it is set.

Reply:
I have used configuration file as the simplest option to open discussion. What about `Gem.data_home`?

https://github.com/rubygems/rubygems/blob/41ce5ce11d0e76a964a8e8a43501a7e389fbf58d/lib/rubygems/defaults.rb#L146-L148

We can create separate file like `last_update_check` and fill it simply with the timestamp number.

Reply:
Alternatively `Gem.cache_home` respecting `XDG_CACHE_HOME` could be the choice.



> I believe `Gem.configuration` already respects `XDG_DATA_HOME` if it is set.

If I understand it well, Gem.configuration respects `XDG_CONFIG_HOME` and that's probably the subject to put into version control.

Reply:
> If I understand it well, `Gem.configuration` respects `XDG_CONFIG_HOME` and that's probably the subject to put into version control.

Yep, exactly! Either `cache_home` or `data_home` would be fine with me, as long as its outside of `config_home` 😄

Reply:
:thinking: Looking at https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html again, it seems `XDG_STATE_HOME` is directory we're looking for.

> $XDG_STATE_HOME defines the base directory relative to which user-specific state files should be stored. If $XDG_STATE_HOME is either not set or empty, a default equal to $HOME/.local/state should be used.

> The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain:
> - actions history (logs, history, recently used files, …)
> - current state of the application that can be reused on a restart (view, layout, open files, undo history, …)

:information_source: I have raised question at freedesktop IRC channel to confirm. :pray: Once done, I'll update this PR.

Reply:
Hi, as the person who brought `XDG_STATE_HOME` into the spec I'd say that folder is the best fit for your file. However, it really is a close call and it putting it into `XDG_DATA_HOME` would be equally fine. So I'd say just pick one of them without overthinking it too much, there is no "wrong" here.

Reply:
Thanks for the suggestion @piegamesde. I had nice chat also at xdg mail list (https://lists.freedesktop.org/archives/xdg/2022-September/014627.html, that's probably where we got your attention) and I decided to go with `XDG_STATE_HOME` here. :muscle:

Reply:
I also agree that `XDG_STATE_HOME` is the best fit!

Reply:
Updated to use `$XDG_STATE_HOME/gem/last_update_check` file with timestamp.